pkg/multiboot/module.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2018 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Module defines modules to be loaded along with the kernel.
+// https://www.gnu.org/software/grub/manual/multiboot/multiboot.html#Boot-information-format.
+
+package multiboot
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/u-root/u-root/pkg/ubinary"
+)
+
+var errEmptyModule = errors.New("empty module")
+
+// A Module represents a module to be loaded along with the kernel.
+type Module struct {
+	// Start is the inclusive start of the Module memory location
+	Start uint32
+	// End is the exclusive end of the Module memory location.
+	End uint32
+
+	// CmdLine is a pointer to a null-terminated ASCII string.
+	CmdLine uint32
+
+	// Reserved is always zero.
+	Reserved uint32
+}
+
+type module struct {
+	start   uint32
+	end     uint32
+	cmdLine string
+}
+
+type modules []module
+
+var sizeofModule = binary.Size(Module{})
+
+func (m modules) size() (uint, error) {
+	b, err := m.marshal(0)
+	return uint(len(b)), err
+}
+
+// marshal writes out the exact bytes of modules to be loaded
+// along with the kernel.
+func (m modules) marshal(base uintptr) ([]byte, error) {
+	dataBuf := bytes.Buffer{}
+	cmdBuf := bytes.Buffer{}
+
+	base += uintptr(sizeofModule * len(m))
+	for _, mod := range m {
+		offset := uint32(base) + uint32(cmdBuf.Len())
+		err := binary.Write(&dataBuf, ubinary.NativeEndian, Module{
+			Start:   mod.start,
+			End:     mod.end,
+			CmdLine: offset,
+		})
+
+		if err != nil {
+			return nil, err
+		}
+
+		if _, err := cmdBuf.WriteString(mod.cmdLine); err != nil {
+			return nil, err
+		}
+
+		if err := cmdBuf.WriteByte(0); err != nil {
+			return nil, err
+		}
+	}
+	data := append(dataBuf.Bytes(), cmdBuf.Bytes()...)
+	return data, nil
+}
+
+func (m *Multiboot) addModules() (uintptr, error) {
+	var modules modules
```

Comment:
var moduleSegs modules?

confusing because there's both modules and m.modules.

Reply:
modules -> moduleDesc